runtime/src/accounts_index.rs

```diff
@@ -581,19 +879,67 @@ impl<T: 'static + Clone> AccountsIndex<T> {
         max_root
     }
 
+    pub fn update_secondary_indexes(
+        &self,
+        pubkey: &Pubkey,
+        slot: Slot,
+        account_owner: &Pubkey,
+        account_data: &[u8],
+        account_indexes: &HashSet<AccountIndex>,
+    ) {
+        if account_indexes.contains(&AccountIndex::ProgramId) {
+            self.program_id_index.insert(account_owner, pubkey, slot);
+        }
+        // Note because of the below check below on the account data length, when an
+        // account hits zero lamports and is reset to Account::Default, then we skip
+        // the below updates to the secondary indexes.
+        //
+        // Skipping means not updating secondary index to mark the account as missing.
+        // This doesn't introduce false positives during a scan because the caller to scan
+        // provides the ancestors to check. So even if a zero-lamport account is not yet
+        // removed from the secondary index, the scan function will:
+        // 1) consult the primary index via `get(&pubkey, Some(ancestors), max_root)`
+        // and find the zero-lamport version
+        // 2) When the fetch from storage occurs, it will return nothing because the account
+        // is default in that storage entry
```

Comment:
Well, could you double check? As far as I re-read the code lightly, the following would be more accurate:


```rust
// 2) When the fetch from storage occurs, it will return Account::Default (as persisted tombstone for snapshots). But then
// it'll ultimately filtered out by post-process redundant filters like at get_filtered_spl_token_accounts_by_owner.
```

Reply:
ah yeah, you're right, for some reason I thought `load` in AccountsDb would return `None` for zero-lamport accounts